scylla compat

clockfort opened the pull request:
- some flexibility for CQL field names
- some resiliancy for CQL errors that the datastax driver doesn't do a good job of catching (an error occurs, the connection closes, and the datastax client driver lets you check out the bad connection from the pool again without checking)
- allow instantiation of multiple hikari pools (this was so I could do some parallel testing, but this is just a good feature)
- add some useful debug logging

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/771)
<!-- Reviewable:end -->

clockfort commented:
@tmgordeeva your change seems good, I have tested it.

clockfort commented:
I found from testing that Tanya's new batching thing does have a bug that people wouldn't hit in practice, which is that batches with extra conditionals (read: put unless exist version of put) can't have the batch span across partition key token boundaries.
This wouldn't happen from people inserting one transaction commit at a time (the only use of PUE currently)

gsheasby commented:
@clockfort do we want to actually merge anything from this PR, given that making "AtlasDB + scylla" a thing is far in the distance?

tpetracca commented:
@gsheasby Per internal thread yes.

rhero commented:
Additional context is this PR contains a bunch of CQLKVS improvements which we want irrespective of Scylla. We'd like to move to a Cassandra 3.X/3.0.X version and also a supported client (CQL) before investing in anything scylla specific.

clockfort commented:
I can split off the two or three scylla-specific parts of the change into another commit, if it pleases the court.

clockfort commented:
>_> baseline made almost all of the files merge conflict

gsheasby commented:
Reviewed 15 of 16 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[atlasdb-cassandra-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceTest.java, line 127 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0dJNm19R2DXYfkK9y:-KQ0dJNnJlJJ9qIPMFq2:brbyggx) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceTest.java#L127)):*
> ```Java
>         } else if (keyValueService instanceof TableSplittingKeyValueService) { // scylla tests
>             KeyValueService delegate = ((TableSplittingKeyValueService) keyValueService).getDelegate(testTable);
>             assert(delegate instanceof CassandraKeyValueService);
> ```

Don't use this kind of assert. We prefer something like Matcher.assertThat, which allows a message to be given.

Also, `keyValueService` is always an instance of CassandraKeyValueService (I assume there was some fancy scylla stuff which has since been removed).

---

*[atlasdb-cassandra-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceTest.java, line 140 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0dfMRkeBIbDfiQ7b1:-KQ0dfMStuX4EHXj8497:bqvhfhm) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceTest.java#L140)):*
> ```Java
>         CfDef clusterSideCf = Iterables.getOnlyElement(knownCfs.stream().filter(cf -> cf.getName().equals("ns__never_seen")).collect(Collectors.toList()));
> 
>         assert(CassandraKeyValueServices.isMatchingCf(kvs.getCfForTable(testTable, tableMetadata), clusterSideCf));
> ```

Again, Assert.assertThat or Matcher.assertThat here.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java, line 376 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0eRDEI0O_8iCRCxG9:-KQ0eRDEI0O_8iCRCxGA:b-bdu0jl) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java#L376)):*
> ```Java
>     }
> 
>     private InetSocketAddress getHostPortForHost(String host) throws UnknownHostException {
> ```

General readability note: it's preferable to put a private method /after/ its first usage (contrary to C++ convention), so that someone reading the file top-down already has context.

Also, would `getAddressForHost` also be accurate?

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java, line 402 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0fQ4O-o-_GlNM6gH8:-KQ0fQ4PQqv3YapwCEof:bxqowsw) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java#L402)):*
> ```Java
>             } else { // normal case, large cluster with many vnodes
>                 for (TokenRange tokenRange : tokenRanges) {
>                     List<InetSocketAddress> hosts = Lists.transform(tokenRange.getEndpoints(), new Function<String, InetSocketAddress>() {
> ```

BTW - this gives ugly greyed-out code in idea. Would prefer the lambda version. If having the try/catch block inside a lambda makes you uncomfortable, you're free to extract it into its own method.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java, line 203 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0gJth-2Xez2y9Ay2g:-KQ0gJth-2Xez2y9Ay2h:b-7f0h73) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java#L203)):*
> ```Java
>         clientPool.runOneTimeStartupChecks();
>         if (configManager.getConfig().scyllaDB() && !configManager.getConfig().safetyDisabled()) {
>             throw new IllegalArgumentException("Not currently allowing Thrift-based access to ScyllaDB clusters; " +
> ```

nit: possibly an IllegalStateException?

I would also like to see blank lines before and after the `if` block.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java, line 206 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0goqVdR9nP33ZHgWC:-KQ0goqVdR9nP33ZHgWD:b-7ycnv6) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java#L206)):*
> ```Java
>                     "there appears to be from our tests an existing correctness bug with semi-complex column selections");
>         }
>         boolean supportsCas = !configManager.getConfig().scyllaDB() && clientPool.runWithRetry(CassandraVerifier.underlyingCassandraClusterSupportsCASOperations);
> ```

This line is too long, and fails checkstyle.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java, line 287 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0iMyfoROWirEBlp-B:-KQ0iMyfoROWirEBlp-C:b-3yadkx) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java#L287)):*
> ```Java
>     // also because compression_options after serialization / deserialization comes back as blank
>     // for the ones we set 4K chunk on... ?!
>     public static boolean isMatchingCf(CfDef clientSide, CfDef clusterSide) {
> ```

Are there tests for this method?

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java, line 291 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0hUSwDi6Gzztom990:-KQ0hUSwDi6Gzztom991:bhm1jdy) ([raw file](https://github.com/palantir/atlasdb/blob/863684d89283eab624ffee4444124e107cb79d85/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java#L291)):*
> ```Java
>         if (!Objects.equal(clientSide.compaction_strategy_options, clusterSide.compaction_strategy_options)) {
>             log.debug("Found client/server disagreement on compaction strategy options for {}."
>                     + " (client = ({}), server = ({}))",
> ```

Mr. Style (first name "Check") says to revert these changes.

Also, there's a lot of duplication here. You could extract a method that takes the option ("compaction strategy options" in this case), the client value, and the server value, and use it six times. Then you'd only need to change "debug" to "info" once! DRY FTW.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java, line 319 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0iB6eEGnYGBnMFd41:-KQ0iB6eEGnYGBnMFd42:bfqs07t) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java#L319)):*
> ```Java
>         }
>         if (!Objects.equal(clientSide.compaction_strategy, clusterSide.compaction_strategy)) {
>             // consider equal "com.whatever.LevelledCompactionStrategy" and "LevelledCompactionStrategy". This String-based API stinks.
> ```

Should use "ends with" rather than "contains", just in case some wise guy sets the compaction strategy to "cassandra" or something.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java, line 322 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0hxDxfLivMWZzuwSi:-KQ0hxDyPq1HFgEVPdMm:bqcjwtp) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java#L322)):*
> ```Java
>             if (clientSide.compaction_strategy != null && clusterSide.compaction_strategy != null &&
>                     !(clientSide.compaction_strategy.contains(clientSide.compaction_strategy) || clusterSide.compaction_strategy.contains(clientSide.compaction_strategy))) {
> 		    log.info("Found client/server disagreement on compaction_strategy for {}. (client = ({}), server = ({}))",
> ```

We use spaces and not tabs.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 136 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0mMQHgL_HqKdfDxON:-KQ0mMQHgL_HqKdfDxOO:b-74n2ls) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L136)):*
> ```Java
>     private final CassandraKeyValueServiceConfigManager configManager;
>     private final Optional<CassandraJmxCompactionManager> compactionManager;
>     protected final CQLFieldNameProvider fieldName;
> ```

I would prefer to see `protected` and `private` fields in separate blocks.

The mixing of package/protected/private/final fields in this class is generally haphazard.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 358 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0noHN0T4qKczluMoE:-KQ0noHN0T4qKczluMoF:bdtcgif) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L358)):*
> ```Java
>     private Map<Cell, Value> getRowsAllColsInternal(final TableReference tableRef,
>                                                     final Iterable<byte[]> rows,
>                                                     final long startTs) throws Exception {
> ```

This method is generally hard to read because:
1 - it's very long
2 - it contains many nested for loops
3 - it doesn't contain a single blank line to separate different blocks/sections 

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 366 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0nP-sB9l_dHyeWVHE:-KQ0nP-sB9l_dHyeWVHF:bfb4hy8) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L366)):*
> ```Java
>             rowCount += batch.size();
>             List<ResultSetFuture> resultSetFutures = Lists.newArrayListWithExpectedSize(rowCount);
>             String getRowsQuery = "SELECT * FROM " + getFullTableName(tableRef) + " WHERE " + fieldName.row() + " IN ("
> ```

nit: Lots of concatenation here. Consider using String.format

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 369 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0p4uX5s65p1oTlX5h:-KQ0p4uX5s65p1oTlX5i:b2r5g5x) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L369)):*
> ```Java
>                     + Joiner.on(",").join(Iterables.limit(Iterables.cycle("?"), batch.size())) + ")";
>             PreparedStatement preparedStatement = getPreparedStatement(tableRef, getRowsQuery, session);
>             Object[] args = new Object[batch.size()];
> ```

Can be replaced with the following Java 8 goodness:
`            Object[] args = batch.stream().map(ByteBuffer::wrap).toArray();`

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 431 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0pfrJnRCeXU0-u-PV:-KQ0pfrJnRCeXU0-u-PW:bs3ha3y) ([raw file](https://github.com/palantir/atlasdb/blob/863684d89283eab624ffee4444124e107cb79d85/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L431)):*
> ```Java
>                             final Visitor<Multimap<Cell, Value>> visitor,
>                             final ConsistencyLevel consistency) throws Exception {
>         String loadWithTsQuery = "SELECT * FROM " + getFullTableName(tableRef)
> ```

Again, this looks like someone optimised once only, although they did so in an inefficient way (see Effective Java, second edition, item 51).

It's probably better to declare the common part outside of the for loop (something like `String query = "SELECT /*snip*/ WHERE {} IN ({}) AND {} = ? /*snip*/"`, and then using `String.format` inside the for loop.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 439 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0pYA7a9SWPDZa3VxR:-KQ0pYA7a9SWPDZa3VxS:bchiad0) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L439)):*
> ```Java
> 
>         List<ResultSetFuture> resultSetFutures = Lists.newArrayListWithCapacity(cells.size());
>         TreeMultimap<byte[], Cell> cellsByCol =
> ```

Baseline forbids the usage of this class.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 462 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0qma2j81j2HDrmzoV:-KQ0qma2j81j2HDrmzoW:be6chjh) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L462)):*
> ```Java
>                               .setConsistencyLevel(consistency);
>                 Object[] args = new Object[batch.size() + 2];
>                 for (int i = 0; i < batch.size(); i++) {
> ```

Oh hey, another opportunity to write some Java 8 goodness (although there's a weird extra two arguments to deal with in this case)

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 529 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0rLYctChl6KEk0pFo:-KQ0rLYctChl6KEk0pFp:b-f7ljif) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L529)):*
> ```Java
>         List<Future<Map<Cell, Long>>> futures = Lists.newArrayListWithCapacity(numPartitions);
>         String loadOnlyTsQuery = "SELECT "
>                 + fieldName.row() + ", "
> ```

I had to look up what `fieldName` was, indicating that it might be better named `fieldNameProvider`.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueServices.java, line 281 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0s3nyzeMKg3gPHW0M:-KQ0s3nyzeMKg3gPHW0N:b6hso3t) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueServices.java#L281)):*
> ```Java
> 
>         queryBuilder.append("CREATE TABLE IF NOT EXISTS " + kvs.getFullTableName(tableRef) + " ( " // full table name (ks.cf)
>                 + kvs.fieldName.row() + " blob, "
> ```

BTW - it's sad that you have to pass in the whole KVS to this method, and then mostly use `fieldName` and a few (well, several).

There are probably classes hiding within CQLKVS. For instance, there's probably one that could contain logic like fieldName, but also how to get table names from table refs (something like CQLSchema?)

The remaining usages of kvs, right at the bottom of this function, are all to do with executing a query. In a parallel PR, George and I built `CqlExecutor`; something similar to this could be responsible for executing the queries, and it sounds reasonable for it to know of `longRunningQuerySession`, `session`, and `cqlStatementCache`.

Note that I don't think you should do any of these refactors as part of this PR; they belong in a separate refactoring PR, once CQLKVS is well-tested.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLStatementCache.java, line 55 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0t286GoN9XsUasiSV:-KQ0t286GoN9XsUasiSW:bfzcthl) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLStatementCache.java#L55)):*
> ```Java
>             return currentSession.prepare(query);
>         } catch (NoHostAvailableException e) {
>             if (currentSession.isClosed()) {
> ```

I can see what you're doing here, but this could do with some explanation.

---

*[atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java, line 292 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0u900j9krb9d1yaBT:-KQ0u900j9krb9d1yaBU:b-p4idoc) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java#L292)):*
> ```Java
> 
>         HikariDataSource dataSourcePool;
>         try {
> ```

This looks exceptionally weird. The method is also too long. You can probably keep the declaration and initialisation of a variable together by extracting the try(x2) block into a method called `getDataSourcePool`.

---

*[atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java, line 299 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0uTwrIw7BRXRsfVVl:-KQ0uTwrIw7BRXRsfVVm:buzmdmf) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java#L299)):*
> ```Java
>                     String poolName = connConfig.getHikariConfig().getPoolName();
>                     connConfig.getHikariConfig().setPoolName((poolName + "-" + ThreadLocalRandom.current().nextInt()));
>                     dataSourcePool = new HikariDataSource(connConfig.getHikariConfig());
> ```

It's also odd (again) that you're doing a thing, catching an unchecked exception, and then retrying the first thing (but without catching the same exception). At least comment that that's what you're doing.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableSplittingKeyValueService.java, line 178 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0utCWr8-3z38Pjcpt:-KQ0utCXsjiN62RMTbXQ:bpsjmj) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableSplittingKeyValueService.java#L178)):*
> ```Java
>     }
> 
>     public KeyValueService getDelegate(TableReference tableRef) {
> ```

If the change in CKVSTest was Scylla-only, then this is also probably Scylla-only.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/771)*
<!-- Sent from Reviewable.io -->

clockfort commented:
Review status: 8 of 16 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[atlasdb-cassandra-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceTest.java, line 127 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0dJNm19R2DXYfkK9y:-KQ2NHbPiqv9Dw5FwIlg:b-v2g1b8) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceTest.java#L127)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Don't use this kind of assert. We prefer something like Matcher.assertThat, which allows a message to be given.
> 
> Also, `keyValueService` is always an instance of CassandraKeyValueService (I assume there was some fancy scylla stuff which has since been removed).
</details>
As the comment says, the scylla tests (and any usage of scylla) will all actually show up as TableSplittingKVSs because they don't support Put Unless Exist and to support them I just split the transaction table to a postgres. I just haven't included the scylla tests because Atlas probably doesn't want tests that aren't automatically run.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java, line 402 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0fQ4O-o-_GlNM6gH8:-KQ2PiylYxd7ROYpKluO:b-88oegz) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java#L402)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> BTW - this gives ugly greyed-out code in idea. Would prefer the lambda version. If having the try/catch block inside a lambda makes you uncomfortable, you're free to extract it into its own method.
</details>
Yep, kept it in the old style because of the try/catch and wanted to avoid the spaghetti code of making tiny methods everywhere to make lambdas look nice.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java, line 287 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0iMyfoROWirEBlp-B:-KQ2OYwajVu_ssv1nSLO:b-ua6ljw) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java#L287)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Are there tests for this method?
</details>
... 
I added a test for it (testCfEqualityChecker) in this CR. You /just/ CR'd it.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 439 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0pYA7a9SWPDZa3VxR:-KQ2_FrWX7vva8_7H6I0:b-vds0ox) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L439)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Baseline forbids the usage of this class.
</details>
Okay, will just call it a SortedSetMultiMap. I think baseline's being shortsighted for the reasons they list for this (concrete coupling). They really should only care about this sort of thing in your method signatures and not your method-local variables. 

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 462 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0qma2j81j2HDrmzoV:-KQ3G7mk5vO_oLWpZ3Nq:b-szylko) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L462)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Oh hey, another opportunity to write some Java 8 goodness (although there's a weird extra two arguments to deal with in this case)
</details>
I'm going to keep this one of Tanya's Java6isms, I don't think it will look any less weird in java8 stream land given those two extras.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/771)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java, line 402 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0fQ4O-o-_GlNM6gH8:-KQ5mZc8MC7sikaS38rn:b-nqcq64) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java#L402)):*
<details><summary><i>Previously, clockfort (Chris Lockfort) wrote…</i></summary>
> Yep, kept it in the old style because of the try/catch and wanted to avoid the spaghetti code of making tiny methods everywhere to make lambdas look nice.
</details>
I'd say that the main problem with this class is that it's 600+ lines long - and therefore it would be easy to lose small methods abstracted out of little parts of the mess. Fixing that, though, is waaaay beyond the scope of this PR.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java, line 206 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0goqVdR9nP33ZHgWC:-KQ5n39khPMCj1GRBTal:b-zepucu) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java#L206)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> This line is too long, and fails checkstyle.
</details>
LGTM now!

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java, line 287 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0iMyfoROWirEBlp-B:-KQ5nLYouPu1NAY_GTTh:btf9er0) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java#L287)):*
<details><summary><i>Previously, clockfort (Chris Lockfort) wrote…</i></summary>
> ... 
> I added a test for it (testCfEqualityChecker) in this CR. You /just/ CR'd it.
</details>
...yes, yes I did. Sorry, lazy comment!

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java, line 291 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0hUSwDi6Gzztom990:-KQ5nW08qJZ6GltAp9jK:b6xapi5) ([raw file](https://github.com/palantir/atlasdb/blob/863684d89283eab624ffee4444124e107cb79d85/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServices.java#L291)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Mr. Style (first name "Check") says to revert these changes.
> 
> Also, there's a lot of duplication here. You could extract a method that takes the option ("compaction strategy options" in this case), the client value, and the server value, and use it six times. Then you'd only need to change "debug" to "info" once! DRY FTW.
</details>
Much nicer, thanks!

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 439 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0pYA7a9SWPDZa3VxR:-KQ5nnNErDV0f-aMKChP:b-pbdqz3) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L439)):*
<details><summary><i>Previously, clockfort (Chris Lockfort) wrote…</i></summary>
> Okay, will just call it a SortedSetMultiMap. I think baseline's being shortsighted for the reasons they list for this (concrete coupling). They really should only care about this sort of thing in your method signatures and not your method-local variables. 
</details>
Agreed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/771)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java, line 203 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0gJth-2Xez2y9Ay2g:-KQ5sHu3wTGTHJRJzXbd:b-xz8qh2) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java#L203)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> nit: possibly an IllegalStateException?
> 
> I would also like to see blank lines before and after the `if` block.
</details>
Thanks for the blanks. Actually thinking about it, we should throw the Illegal?Exception before running startup checks, since the startup checks have the side-effect of also creating the keyspace (!)

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java, line 367 \[r4\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ5soLT2I-EfhQcUJzh:-KQ5soLU1W5qa0M-4kW1:b3pzuwl) ([raw file](https://github.com/palantir/atlasdb/blob/fe7369a82acbf8a11c1f0c439493d2000dac898e/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java#L367)):*
> ```Java
>         for (final List<byte[]> batch : Iterables.partition(rows, fetchBatchCount)) {
>             rowCount += batch.size();
>             List<ResultSetFuture> resultSetFutures = Lists.newArrayListWithExpectedSize(rowCount);
> ```

Is this a bug? For each iteration, we create a new list of resultSetFutures. It looks like only the results of the current batch goes into that list - so the list of ResultSetFutures looks like it will have a constant size (actually, we only ever put one element nito the list). However, the size of the list grows steadily.

By the end of the loop, rowCount will equal the total size of rows, and the warning will be generated any time we go through the loop more than once.

---

*[atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java, line 292 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/771#-KQ0u900j9krb9d1yaBT:-KQ5tnwkyfIPm5Zb4Iey:brs1xae) ([raw file](https://github.com/palantir/atlasdb/blob/576911202e65894b50594f2ad3d5ae247678a2c9/atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java#L292)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> This looks exceptionally weird. The method is also too long. You can probably keep the declaration and initialisation of a variable together by extracting the try(x2) block into a method called `getDataSourcePool`.
</details>
// don't care enough about this one to block it. I assume the second exception can be thrown by ``new HikariDataSource``.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/771)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
@clockfort this is still waiting on a few things. My only blockers are:
1. the question about a potential bug in CQLKeyValueService (line 367)
2. the ordering of statements in CassandraKeyValueService.init() - to be changed to avoid side-effects

Circle is blocking it because the build is red. Probably checkstyle.

rjullman commented:
Reviewed 6 of 17 files at r1, 2 of 8 files at r4, 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/771)*
<!-- Sent from Reviewable.io -->
